src/sentry/integrations/source_code_management/repository.py

```diff
@@ -0,0 +1,200 @@
+        """
+        Calls the client's `check_file` method to see if the file exists.
+        Returns the link to the file if it's exists, otherwise return `None`.
+
+        So far only GitHub, GitLab and VSTS have this implemented, all of which give
+        use back 404s. If for some reason an integration gives back a different
+        status code, this method could be overwritten.
+
+        repo: Repository (object)
+        filepath: file from the stacktrace (string)
+        branch: commitsha or default_branch (string)
+        """
+        filepath = filepath.lstrip("/")
+        try:
+            client = self.get_client()
+        except (Identity.DoesNotExist, IntegrationError):
+            sentry_sdk.capture_exception()
+            return None
+        try:
+            response = client.check_file(repo, filepath, branch)
+            if response is None:
+                return None
+        except IdentityNotValid:
+            return None
+        except ApiError as e:
+            if e.code != 404:
+                sentry_sdk.capture_exception()
+                raise
+
+            return None
+
+        return self.format_source_url(repo, filepath, branch)
+
+    def get_stacktrace_link(
+        self, repo: Repository, filepath: str, default: str, version: str | None
+    ) -> str | None:
+        """
+        Handle formatting and returning back the stack trace link if the client
+        request was successful.
+
+        Uses the version first, and re-tries with the default branch if we 404
+        trying to use the version (commit sha).
+
+        If no file was found return `None`, and re-raise for non-"Not Found"
+        errors, like 403 "Account Suspended".
+        """
+        scope = sentry_sdk.Scope.get_isolation_scope()
+        scope.set_tag("stacktrace_link.tried_version", False)
+        if version:
+            scope.set_tag("stacktrace_link.tried_version", True)
+            source_url = self.check_file(repo, filepath, version)
+            if source_url:
+                scope.set_tag("stacktrace_link.used_version", True)
+                return source_url
+        scope.set_tag("stacktrace_link.used_version", False)
+        source_url = self.check_file(repo, filepath, default)
+
+        return source_url
+
+    @abstractmethod
+    def get_repositories(self, query: str | None = None) -> Sequence[dict[str, Any]]:
+        """
+        Get a list of available repositories for an installation
+
+        >>> def get_repositories(self):
+        >>>     return self.get_client().get_repositories()
+
+        return [{
+            'name': display_name,
+            'identifier': external_repo_id,
+        }]
+
+        The shape of the `identifier` should match the data
+        returned by the integration's
+        IntegrationRepositoryProvider.repository_external_slug()
+        """
+        raise NotImplementedError
+
+    def get_unmigratable_repositories(self) -> Collection[RpcRepository]:
```

Comment:
the default, if unimplemented, is []. again this could possibly be an abstract method and return [] everywhere that doesn't have an overriding implementation

Reply:
👍 but nit: `()` is preferable.